source/common/stats/symbol_table_impl.h

```diff
@@ -39,6 +39,152 @@ constexpr uint64_t StatNameMaxSize = 1 << (8 * StatNameSizeEncodingBytes); // 65
+
+  // Obtains new backing storage for an already existing StatName. Used to
+  // record a computed StatName held in a temp into a more persistent data
+  // structure.
+  StatNameStorage(StatName src, SymbolTable& table);
+
+  /**
+   * Before allowing a StatNameStorage to be destroyed, you must call free()
+   * on it, to drop the references to the symbols, allowing the SymbolTable
+   * to shrink.
+   */
+  ~StatNameStorage();
+
+  /**
+   * Decrements the reference counts in the SymbolTable.
+   *
+   * @param table the symbol table.
+   */
+  void free(SymbolTable& table);
+
+  /**
+   * @return StatName a reference to the owned storage.
+   */
+  inline StatName statName() const;
+
+  uint8_t* bytes() { return bytes_.get(); }
+
+private:
+  SymbolTable::StoragePtr bytes_;
+};
+
+/**
+ * Maintains storage for a collection of StatName objects. Like
+ * StatNameManagedStorage, this has an RAII usage model, taking
+ * care of decrementing ref-counts in the SymbolTable for all
+ * contained StatNames on destruction or on clear();
+ *
+ * Example usage:
+ *   StatNamePool pool(symbol_table);
+ *   StatName name1 = pool.add("name1");
+ *   StatName name2 = pool.add("name2");
+ *   uint8_t* storage = pool.addReturningStorage("name3");
+ *   StatName name3(storage);
+ */
+class StatNamePool {
+public:
+  explicit StatNamePool(SymbolTable& symbol_table) : symbol_table_(symbol_table) {}
+  ~StatNamePool() { clear(); }
+
+  /**
+   * Removes all StatNames from the pool.
+   */
+  void clear();
+
+  /**
+   * @param name the name to add the container.
+   * @return the StatName held in the container for this name.
+   */
+  StatName add(absl::string_view name);
+
+  /**
+   * Does essentially the same thing as add(), but returns the storage as a
+   * pointer which can later be used to create a StatName. This can be used
+   * to accumulate a vector of uint8_t* which can later be used to create
+   * StatName objects on demand.
+   *
+   * The use-case for this is in source/common/http/codes.cc, where we have a
+   * fixed sized array of atomic pointers, indexed by HTTP code. This API
+   * enables storing the allocated stat-name in that array of atomics, which
+   * enables content-avoidance when finding StatNames for frequently used HTTP
+   * codes.
+   *
+   * @param name the name to add the container.
+   * @return a pointer to the bytes held in the container for this name, suitable for
+   *         using to construct a StatName.
+   */
+  uint8_t* addReturningStorage(absl::string_view name);
+
+private:
+  // We keep the stat names in a vector of StatNameStorage, storing the
+  // SymbolTable reference separately. This saves 8 bytes per StatName,
+  // at the cost of having a destructor that calls clear().
+  SymbolTable& symbol_table_;
+  std::vector<StatNameStorage> storage_vector_;
+};
+
+// Represents a map from string to StatName, including the pool used to allocate
+// new StatNames.
+class StatNameMappedPool {
```

Comment:
Note: I owe unit tests for this class. Wanted to start getting feedback on the rest of the PR in parallel.